### DISCUSSION TITLE
Copying Elements

### DISCUSSION THREAD

**Discussion by PoshoDev** [zauberzeug/nicegui#3083] [Q&A] [answered]
### Question

Hello there! :grin:

I am working on a class of sorts that duplicates elements along its inner rows. The thing is, I'm not exactly sure if I'm on the right path. I noticed the `Element` class has no `copy()` method or similar, so I had to improvise a bit.

This is the most minimum reproducible code I was able to put together:

```python
from copy import copy

from nicegui import ui


class MyElement():
    def __init__(self, rows: int):
        self.container = ui.column()
        self.rows = []
        with self.container:
            for _ in range(rows):
                self.rows.append(ui.row().classes('border-[1px]'))
            self.waitingline = ui.column()

    def __enter__(self):
        self.waitingline.__enter__()
    
    def __exit__(self, exc_type, exc_value, exc_traceback):
        children = copy(self.waitingline.slots["default"].children)
        for child in children:
            for row in self.rows:
                new_row = copy(child)
                row.default_slot.children.insert(-1, new_row)
                new_row.parent_slot = row.default_slot
        # self.waitingline.clear()
                
with MyElement(4):
    with ui.row():
        ui.label("Hello")
        ui.label("World")

ui.run()
```
![image](https://github.com/zauberzeug/nicegui/assets/32964490/2f54c815-8e21-4c8d-a4d7-412688e3857c)

That kinda works, but if then I want to clear the `waitingline` (see commented line on code above) then everything from every container disappears:

![image](https://github.com/zauberzeug/nicegui/assets/32964490/486f5a73-3303-4a22-a2f8-0389286db157)

I've tried fiddling in many other ways by taking hints from the `Element` class methods (`move()`, `clear()`, `delete()`, `remove()`) to no avail.

I finally got to get it working the way I want by making the `waitingline` invisible (shown where `# HACK` is commented), but it doesn't feel like a proper solution and I feel it can lead to problems down the line:

```python
from copy import copy

from nicegui import ui


class MyElement():
    def __init__(self, rows: int):
        self.container = ui.column()
        self.rows = []
        with self.container:
            for _ in range(rows):
                self.rows.append(ui.row().classes('border-[1px]'))
            self.waitingline = ui.column()
            self.waitingline.set_visibility(False)  # HACK

    def __enter__(self):
        self.waitingline.set_visibility(True)  # HACK
        self.waitingline.__enter__()
    
    def __exit__(self, exc_type, exc_value, exc_traceback):
        children = copy(self.waitingline.slots["default"].children)
        for child in children:
            for row in self.rows:
                new_row = copy(child)
                row.default_slot.children.insert(-1, new_row)
                new_row.parent_slot = row.default_slot
        # self.waitingline.clear()
        self.waitingline.set_visibility(False)  # HACK

with MyElement(4):
    with ui.row():
        ui.label("Hello")
        ui.label("World")

ui.run()
```
![image](https://github.com/zauberzeug/nicegui/assets/32964490/670f60ab-996e-4c67-aab9-bfcf72afb9c6)

What would be the actual proper way for copying elements? Could anyone lend me a hand with this? :thinking: 

Thanks a lot in advance!

**Comment by falkoschindler** [comment, marked as answer]
Hi @PoshoDev,

Copying UI elements is currently not supported. It's probably complicated to get it right, because it's unclear what to copy and what not (e.g. event handlers, children etc.). We'd suggest to use a data model as a source of truth and then create UI elements based on that model.

**Reply by PoshoDev** [reply]
Yeah, that sounds about what I ended up doing in the end. Would've been nice though.

Thank you very much! :)
